examples/vision/knowledgedistillation.py

```diff
@@ -0,0 +1,224 @@
+        layers.Conv2D(512, (3, 3), strides=(2, 2), padding="same"),
+        layers.Flatten(),
+        layers.Dense(10),
+    ],
+    name="teacher",
+)
+
+# Create the student
+student = keras.Sequential(
+    [
+        keras.Input(shape=(28, 28, 1)),
+        layers.Conv2D(16, (3, 3), strides=(2, 2), padding="same"),
+        layers.LeakyReLU(alpha=0.2),
+        layers.MaxPooling2D(pool_size=(2, 2), strides=(1, 1), padding="same"),
+        layers.Conv2D(32, (3, 3), strides=(2, 2), padding="same"),
+        layers.Flatten(),
+        layers.Dense(10),
+    ],
+    name="student",
+)
+
+"""
+## Prepare dataset
+The dataset used for training the teacher and distilling the teacher is
+[MNIST](https://keras.io/api/datasets/mnist/). Both the student and teacher are trained
+on the training set and evaluated on the test set.
+"""
+
+# Prepare the train and test dataset.
+batch_size = 64
+(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
+
+# Normalize data
+x_train = x_train.astype("float32") / 255.0
+x_train = np.reshape(x_train, (-1, 28, 28, 1))
+
+x_test = x_test.astype("float32") / 255.0
+x_test = np.reshape(x_test, (-1, 28, 28, 1))
+
+# Construct datasets
+train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+train_dataset = train_dataset.shuffle(buffer_size=1024).batch(batch_size)
+
+test_dataset = tf.data.Dataset.from_tensor_slices((x_test, y_test))
+test_dataset = test_dataset.shuffle(buffer_size=1024).batch(batch_size)
+
+"""
+## Train teacher
+In knowledge distillation we assume that the teacher is trained and fixed. Thus, we start
+by training the teacher model on the training set in the usual way.
+"""
+
+# Train teacher as usual
+teacher.compile(
+    optimizer="adam",
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    metrics=["accuracy"],
+)
+
+# Train and evaluate on subset of data to reduce computations.
+teacher.fit(train_dataset.take(10), epochs=1)
+loss, accuracy = teacher.evaluate(test_dataset.take(10))
+
+"""
+## Distill teacher to student
+We have already trained the teacher model, and we only need to initialize a
+`Distiller(student, teacher)` instance, `compile()` it with the desired losses,
+hyperparameters and optimizer, and distill the teacher to the student.
+"""
+
+# Initialize and compile distiller
+distiller = Distiller(student=student, teacher=teacher)
+distiller.compile(
+    optimizer="adam",
+    metrics=["accuracy"],
+    s_loss_fn=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    d_loss_fn=keras.losses.KLDivergence(),
+    alpha=0.1,
+    T=3,
+)
+
+# Distill teacher to student
+distiller.fit(train_dataset.take(10), epochs=1)
+
+# Evaluate student on test dataset
+accuracy = distiller.evaluate(test_dataset.take(10))
+
+"""
+If the teacher is trained for 5 full epochs and the student is distilled on this teacher
+for 5 epochs, you should experience a performance boost compared to training the same
```

Comment:
Is this the case with this example? What are the numbers?

Reply:
I have added an additional section, where the student model is trained from scratch without knowledge distillation, along with performance metrics when training on all data. With a fixed seed, it amounts to approx.:

- Teacher: 97.6%
- Distillation student: 98.1%
- Classical student: 97.6%